cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAutoRebalancingReconciler.java

```diff
@@ -0,0 +1,530 @@
+import io.strimzi.operator.cluster.operator.resource.ResourceOperatorSupplier;
+import io.strimzi.operator.cluster.operator.resource.kubernetes.CrdOperator;
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.ReconciliationLogger;
+import io.strimzi.operator.common.model.Labels;
+import io.strimzi.operator.common.model.StatusUtils;
+import io.vertx.core.Future;
+
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static io.strimzi.api.ResourceAnnotations.ANNO_STRIMZI_IO_REBALANCE;
+import static io.strimzi.api.ResourceAnnotations.ANNO_STRIMZI_IO_REBALANCE_AUTOAPPROVAL;
+
+/**
+ * This class runs the reconciliation for the auto-rebalancing process when the Kafka cluster is scaled up/down.
+ */
+public class KafkaAutoRebalancingReconciler {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaAutoRebalancingReconciler.class.getName());
+
+    private static final String STRIMZI_IO_AUTO_REBALANCING_FINALIZER = "strimzi.io/auto-rebalancing";
+
+    private static final ScalingNodes EMPTY_SCALING_NODES = new ScalingNodes(Set.of(), Set.of());
+
+    private final Reconciliation reconciliation;
+    private final Kafka kafkaCr;
+    private final CrdOperator<KubernetesClient, KafkaRebalance, KafkaRebalanceList> kafkaRebalanceOperator;
+    private final Set<Integer> scalingDownBlockedNodes;
+    private final Set<Integer> scalingUpAddedNodes;
+
+    /**
+     * Constructs the Kafka auto-rebalancing reconciler
+     *
+     * @param reconciliation    Reconciliation marker
+     * @param kafkaCr   The Kafka custom resource
+     * @param supplier  Supplies the operators for different resources
+     * @param scalingDownBlockedNodes  nodes blocked on scaling down because of the need for auto-rebalancing first
+     */
+    public KafkaAutoRebalancingReconciler(
+            Reconciliation reconciliation,
+            Kafka kafkaCr,
+            ResourceOperatorSupplier supplier,
+            Set<Integer> scalingDownBlockedNodes) {
+        this.reconciliation = reconciliation;
+        this.kafkaCr = kafkaCr;
+        this.kafkaRebalanceOperator = supplier.kafkaRebalanceOperator;
+        this.scalingDownBlockedNodes = scalingDownBlockedNodes;
+        this.scalingUpAddedNodes = new LinkedHashSet<>();
+    }
+
+    /**
+     * The main reconciliation method which triggers the whole reconciliation pipeline. This is the method which is
+     * expected to be called from the outside to trigger the reconciliation.
+     *
+     * @param kafkaStatus The Kafka Status class for updating auto-rebalancing status on it during the reconciliation
+     *
+     * @return  Future which completes when the reconciliation completes
+     */
+    public Future<Void> reconcile(KafkaStatus kafkaStatus) {
```

Comment:
> not store the whole Kafka CR

It was also used across the reconciler to get the namespace (which actually I can get from the reconciliation marker I already have) and the cluster name for generating the KafkaRebalance custom resource. Anyway I can just store the cluster name in a class field. Of course, the `KafkaRebalanceStatus` can be stored in the ctor in a class field and then used later as well as ScalingNodes record. I will take a look at this refactoring.

Reply:
Actually the cluster name is in the reconciliation marker as well :-)

Reply:
So I have refactored the code so that Kafka CR is not needed to be stored anymore. The ctor just get what the class needs from that (the autorebalance status and configuration). Regarding the `ScalingNodes` record and corresponding `getScalingNodes` method it's still needed in the reconcile because the added nodes are coming from the `KafkaStatus` instance (filled by the `KafkaReconciler`). So instead of setting a field member in the reconcile about added nodes, I think having a record helps readability and makes it clear on what the computation of next status is going to work. Related to that I know that the `kafkaAutoRebalanceStatus` is now a class member (so methods like `computeNextStatus` and the various `onXXXX` could just access it) so I could avoid to pass but my feeling is that that readability wise it's still better. But I have still not strong opinion on it, so happy to change if any good reasons.

Reply:
How does it help the readablity to have two different things called `kafkaAutoRebalanceStatus` -> one as a class member and one as passed through a method parameter? Doesn't that just create confusion which is which and where should be used which one?

Reply:
I am fine changing it and removing such parameters to use the class member everywhere.